src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        // 1 - node - 3
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+        if (node.next == null) {
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        int count = 0;
+        Node<T> pointer = head;
+        while (pointer != null) {
+            if (count == index) {
+                return pointer;
+            }
+            pointer = pointer.next;
+            count++;
+        }
+        return null;
+    }
+
+    private void testIndex(int index) {
+        if (index >= size() || index < 0) {
+            throw new IndexOutOfBoundsException("index outside size: " + index);
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNoge = new Node<>(tail, null, value);
+        if (head == null) {
+            head = newNoge;
+        }
+        if (tail != null) {
+            tail.next = newNoge;
+        }
+        tail = newNoge;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size()) {
+            add(value);
+            return;
+        }
+        testIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, head, value);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, currentNode, value);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        testIndex(index);
+        Node<T> pointer = getNode(index);
+        if (pointer == null) {
+            return null;
+        } else {
+            return pointer.getValue();
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        testIndex(index);
+        int count = 0;
+        Node<T> pointer = head;
+        T oldValue = null;
+        while (pointer != null) {
+            if (index == count) {
+                oldValue = pointer.getValue();
+                pointer.setValue(value);
+                break;
+            } else {
+                pointer = pointer.next;
+                count++;
+            }
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        testIndex(index);
+        T toReturn = null;
+        Node<T> pointer = getNode(index);
+        if (pointer != null) {
+            T result = pointer.getValue();
+            unlink(pointer);
+            size--;
```

Comment:
better include size-- to unlink method instead of repeatedly calling them together